_posts/2023-04-02-writing-custom-pod-security-providers.md

```diff
@@ -0,0 +1,269 @@
+This interface contains several different methods which will need to be implemented in your code.
+In this example, we will implement a provider which will configure all containers to use a read-only root filesystem.
+Read-only root filesystem means that the filesystem of the container image will be always read-only and the containers will be able to write only to folders mounted as volumes in the Pod definition.
+
+To get started, we create a new class `CustomPodSecurityProvider` and let it implement the `PodSecurityProvider`
+
+```java
+package cz.scholz.providers;
+
+// Imports
+
+public class CustomPodSecurityProvider implements PodSecurityProvider {
+    // Implemented interface methods
+}
+```
+
+The first method is the `configure(...)` method which is called when the provider is loaded and can be used to configure and initialize the provider.
+This method consumes as a single parameter a [`PlatformFeatures` object](https://github.com/strimzi/strimzi-kafka-operator/blob/main/api/src/main/java/io/strimzi/platform/PlatformFeatures.java).
+You can use this object to find out more about the environment in which the operator is running - in particular the Kubernetes version.
+You can use this if the security context should be set differently for different Kubernetes versions.
+In many cases, you will not have anything to configure.
+In that case, you can just _do nothing_ in this method:
+
+```java
+    @Override
+    public void configure(PlatformFeatures platformFeatures) {
+        // Nothing to configure
+    }
+```
+
+Next, the interface defines several methods which create the Pod Security Context and (container) Security Context.
+These methods will exist for every Pod and every container created by Strimzi.
+For example, the interface defines the following methods for the Kafka pods:
+* `PodSecurityContext kafkaPodSecurityContext(PodSecurityProviderContext context)` for the Pod Security Context
+* `SecurityContext kafkaContainerSecurityContext(ContainerSecurityProviderContext context)` for the Security Context of the init container used for rack awareness or node port listeners
+* `SecurityContext kafkaInitContainerSecurityContext(ContainerSecurityProviderContext context)` for the security context of the main Apache Kafka container
+
+When these methods are called by Strimzi, they will always get an object of type `PodSecurityProviderContext` or `ContainerSecurityProviderContext` as a parameter.
+the context contains two types of information:
+* The storage configuration that might be needed to set some specific configuration for the stateful operands concerning their storage.
+  We do not care about the storage in this example, so we can ignore it.
```

Comment:
```suggestion
1. The first is storage configuration, which may be needed to configure specific storage settings for stateful operands. 
However, since this example focuses on security context, we can ignore storage configuration.
```